File: rudolph/src/main/java/cn/wzbos/android/rudolph/router/Router.java
package cn.wzbos.android.rudolph.router;


import android.content.Context;
import android.net.Uri;
import android.os.Bundle;
import android.text.TextUtils;

import java.util.List;
import java.util.Map;

import cn.wzbos.android.rudolph.Interceptor;
import cn.wzbos.android.rudolph.OnRouteListener;
import cn.wzbos.android.rudolph.RouteType;
import cn.wzbos.android.rudolph.Rudolph;

public abstract class Router<T> {

    String rawUrl;
    Bundle bundle;
    Class<?> target;
    OnRouteListener callback;
    String routePath;
    RouteType routeType;
    String routeTag;
    Map<String, String> queryParameters;

    public String getRawUrl() {
        return rawUrl;
    }

    public Bundle getBundle() {
        return bundle;
    }

    public Class<?> getTarget() {
        return target;
    }

    public String getRoutePath() {
        return routePath;
    }

    public RouteType getRouteType() {
        return routeType;
    }

    public String getRouteTag() {
        return routeTag;
    }

    public Map<String, String> getQueryParameters() {
        return queryParameters;
    }

    public Uri getUriData() {
        if (rawUrl.contains("://")) {
            return Uri.parse(rawUrl);
        }

        if(!TextUtils.isEmpty(Rudolph.getScheme())){
            return Uri.parse(Rudolph.getScheme() + "://" + rawUrl);
        }

        return Uri.EMPTY;
    }

    Router(RouteBuilder<?,?> builder) {
        this.rawUrl = builder.rawUrl;
        this.callback = builder.callback;
        this.bundle = builder.args;
        this.target = builder.target;
        this.routePath = builder.routePath;
        this.routeType = builder.routeType;
        this.routeTag = builder.routeTag;
        this.queryParameters = builder.queryParameters;
    }

    boolean intercept(Context context) {

        List<Interceptor> interceptors = Rudolph.getInterceptors();
        if (null != interceptors && interceptors.size() > 0) {
            for (Interceptor interceptor : interceptors) {
                if (interceptor.intercept(context, this)) {
                    return true;
                }
            }
        }

        return false;
    }

    public abstract T open();
}

File: rudolph/src/main/java/cn/wzbos/android/rudolph/router/Router.kt
package cn.wzbos.android.rudolph.router

import android.app.Activity
import android.app.Application
import android.content.Context
import android.net.Uri
import android.os.Bundle
import androidx.fragment.app.Fragment
import cn.wzbos.android.rudolph.*
import cn.wzbos.android.rudolph.logger.RLog

abstract class Router<T> internal constructor(builder: RouteBuilder<*, *>) :
    IRouter<Router<T>, T, Bundle> {
    //调用地址
    override var rawUrl: String = builder.rawUrl
    override var extras: Bundle? = builder.extras
    override var target: Class<*>? = builder.target
    override var callback: IRouter.Callback? = builder.callback
    override var routeSchema: MutableList<String>? = builder.scheme
    override var routeHost: MutableList<String>? = builder.host
    override var routePath: MutableList<String>? = builder.path
    override var routeType: RouteType? = builder.routeType
    override var routeTag: String? = builder.routeTag
    override var extraTypes: MutableMap<String, ExtraType>? = builder.extraTypes
    override var interceptors: MutableList<Class<out RouteInterceptor>>? = builder.interceptors
    override var caller: Any? = null
    override fun transform(router: Router<T>) {
        this.rawUrl = router.rawUrl
        this.extras = router.extras
        this.target = router.target
        this.routeSchema = router.routeSchema
        this.routeHost = router.routeHost
        this.routePath = router.routePath
        this.routeType = router.routeType
        this.routeTag = router.routeTag
        this.extraTypes = router.extraTypes
    }

    val context: Context?
        get() {
            return try {
                when (caller) {
                    is Fragment -> (caller as? Fragment)?.activity
                    is Activity -> (caller as? Activity)
                    is Application -> (caller as? Context)
                    else -> Rudolph.context
                } ?: Rudolph.context
            } catch (e: Exception) {
                RLog.e("Rudolph", "getContext error!", e)
                Rudolph.context
            }
        }

    val uriData: Uri
        get() {
            if (rawUrl.contains("://")) {
                return Uri.parse(rawUrl)
            }
            if (!Rudolph.scheme.isNullOrBlank()) {
                return Uri.parse(Rudolph.scheme + "://" + rawUrl)
            }
            return Uri.EMPTY
        }

    fun intercept(caller: Any?): Boolean {
        this.caller = caller

        interceptors?.forEach {
            if (it.newInstance().intercept(this)) {
                return true
            }
        }

        Rudolph.globalInterceptors?.forEach {
            if (it.intercept(this)) {
                return true
            }
        }
        return false
    }

    @Deprecated(message = "use execute", replaceWith = ReplaceWith("execute()"))
    fun open(): T {
        return execute()
    }

    override fun toString(): String {
        return "Router(rawUrl=$rawUrl, extras=$extras, target=$target, callback=$callback, routeType=$routeType, routeTag=$routeTag, extraTypes=$extraTypes)"
    }


}